### PR TITLE
Modularize aten parameter parser and checker

### PR DESCRIPTION
**Pull request by EikanWang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125308

In this PR, we abstracted the different types of aten operation parameters as `ParameterMetadata`. This structure intends to be used to represent and store the metadata of each aten operation parameter. Currently, it only supports `Tensor`, `TensorList`, and `Scalar`.

```C++
using ParameterMetadataValue = std::variant<TensorMetadata, std::vector<TensorMetadata>, c10::Scalar>;
```

With this PR, we can extend other parameter-type support in a more modularize way, like `string`, `int`, `double`.

cc @ezyang @msaroufim @bdhirsh @anijain2305 @chauhang @voznesenskym @penguinwu @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

Differential Revision: [D59399546](https://our.internmc.facebook.com/intern/diff/D59399546)

### CODE REVIEW

File: torch/csrc/inductor/aoti_eager/kernel_holder.h [hunk 1]
```diff
@@ -14,9 +14,33 @@
...
+  // Represent all the parameters of AOTI kernel
+  std::vector<ParameterMetadata> parameter_metadata_list_;
+  // AOTI model runner to run the AOTI kernel
   std::shared_ptr<AOTIModelContainerRunner> kernel_runner_;
-  std::vector<torch::dynamo::TensorCheck> tensor_checks_;
+  AOTIKernelMetadata() : parameter_metadata_list_(), kernel_runner_(nullptr) {}
+
+  // Check whether the given parameter metadata list is the same as the
+  // parameter metadata list of the AOTI kernel.
+  bool check(
```
Comment (jgong5) [inline torch/csrc/inductor/aoti_eager/kernel_holder.h:28]: `operator==`?

**Review by jgong5 — APPROVED** [review]

File: torch/csrc/inductor/aoti_eager/kernel_holder.h (hunk 1 above)
Comment (EikanWang) [reply torch/csrc/inductor/aoti_eager/kernel_holder.h:28]: It only checks its `parameter_metadata_list_` member and bypasses the `kernel_runer_` check. `operator==` may not exactly match the semantic.

File: torch/_inductor/utils.py [hunk 2]
```diff
@@ -1599,10 +1599,20 @@ def aoti_compile_with_persistent_cache(
...
+        isinstance(input, (supported_scalar_types, torch.Tensor, list))
         for input in flattened_inputs
     ):
-        raise NotImplementedError("Only support tensor, int, float, bool for now")
+        raise NotImplementedError(
+            "Only support tensor, tensor list, int, float, bool for now"
+        )
+
+    for input in flattened_inputs:
+        if isinstance(input, list) and not all(
```
Comment (jansel) [inline torch/_inductor/utils.py:1610]: I'm a bit confused as to why `flattened_inputs` contains lists.  Shouldn't the call to `pytree` "flatten" all the lists away (hence the variable name)?

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/utils.py (hunk 2 above)
Comment (EikanWang) [reply torch/_inductor/utils.py:1653]: Sorry for the late response. Refined the code by avoiding flattening all arguments.

**Comment by EikanWang** [comment]
@jansel , may I know if I have addressed your comments?

**Review by jansel — APPROVED** [review]

**Comment by fbgheith** [comment]
@pytorchbot revert -m "test failures when run by meta" -c ghfirst

Tests fail when run inside Meta.

**Comment by fbgheith** [comment]
Output log:

======================================================================
ERROR: test_aoti_eager_support_out_cuda (caffe2.test.inductor.test_torchinductor.GPUTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/dev/shm/uid-30083/61d46b56-seed-nspid4026547936_cgpid5509235-ns-4026548007/caffe2/test/inductor/test_torchinductor.py", line 10510, in new_test
    return value(self)
  File "/dev/shm/uid-30083/61d46b56-seed-nspid4026547936_cgpid5509235-ns-4026548007/caffe2/test/inductor/test_torchinductor.py", line 806, in test_aoti_eager_support_out
    res_tensor = torch.clamp(
RuntimeError: Error in dlopen: /re_tmp/tmpcn01etqu/aoti_eager/aten/cuda/lib/csrfzcw7r3e6iffordcaijz5agbanf4w3dkyjddtlry77enma5e5/chht2tsdgj2cel44r4frh7okj6fmsm7udjs2d6iyu2b5tsg7wcuw.so: undefined symbol: aoti_torch_device_type_cpu

======================================================================
ERROR: test_aoti_eager_with_scalar_cuda (caffe2.test.inductor.test_torchinductor.GPUTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/dev/shm/uid-30083/61d46b56-seed-nspid4026547936_cgpid5509235-ns-4026548007/torch/testing/_comparison.py", line 1226, in not_close_error_metas
    pair.compare()
  File "/dev/shm/uid-30083/61d46b56-seed-nspid4026547936_cgpid5509235-ns-4026548007/torch/testing/_comparison.py", line 708, in compare
    self._compare_values(actual, expected)
  File "/dev/shm/uid-30083/61d46b56-seed-nspid4026547936_cgpid5509235-ns-4026548007/torch/testing/_comparison.py", line 835, in _compare_values
    compare_fn(
  File "/dev/shm/uid-30083/61d46b56-seed-nspid4026547936_cgpid5509235-ns-4026548007/torch/testing/_comparison.py", line 1014, in _compare_regular_values_close
    matches = torch.isclose(
RuntimeError: Error in dlopen: /re_tmp/tmpe8ddkp51/aoti_eager/aten/cuda/lib/c45k6o4iha4bbbscgaq7t2fd6caoqtits2gxv7jnufcgenkvpvl6/c75w5n4js4ftjoqedcx5p4tgwosbolgt7xqlisqqvzexwvglnq2j.so: undefined symbol: aoti_torch_device_type_cpu

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/dev/shm/uid-30083/61d46b56-seed-nspid4026547936_cgpid5509235-ns-4026548007/caffe2/test/inductor/test_torchinductor.py", line 10510, in new_test
    return value(self)
  File "/dev/shm/uid-30083/61d46b56-seed-nspid4026547936_cgpid5509235-ns-4026548007/caffe2/test/inductor/test_torchinductor.py", line 976, in test_aoti_eager_with_scalar
    self.assertEqual(ref_values, res_values)
  File "/dev/shm/uid-30083/61d46b56-seed-nspid4026547936_cgpid5509235-ns-4026548007/torch/testing/_internal/common_utils.py", line 3592, in assertEqual
    error_metas = not_close_error_metas(
  File "/dev/shm/uid-30083/61d46b56-seed-nspid4026547936_cgpid5509235-ns-4026548007/torch/testing/_comparison.py", line 1232, in not_close_error_metas
    raise RuntimeError(
RuntimeError: Comparing

TensorOrArrayPair(
    id=(0,),
    actual=tensor([-9.8341e-01,  2.2043e+00, -8.5609e-01,  1.0821e-01, -1.5677e+00,
         1.2017e+00, -3.2008e-01, -5.4052e-01, -2.4166e-01,  1.4987e+00,
         9.0678e-01, -2.5401e+00,  1.0601e+00,  4.9114e-01,  9.3326e-01,
        -2.0346e+00,  9.1439e-04, -1.6040e+00,  2.1894e+00,  1.0840e+00,
         2.8241e+00, -8.8914e-01, -6.2965e-01,  7.6015e-01, -2.6187e-01,
        -1.0823e+00, -1.9986e+00,  4.2247e-01, -1.9694e+00,  1.9601e-01,
        -1.4765e+00,  8.4396e-01,  3.0701e-01,  1.5813e-01, -8.6276e-01,
         1.2773e+00, -7.9689e-01,  3.2489e+00, -6.1513e-02,  3.4948e+00,
        -1.1431e+00, -1.7335e+00,  7.7941e-01, -1.6171e+00, -2.1219e+00,
         1.4460e+00,  3.0477e+00, -5.2305e-01, -3.3852e-01,  3.3611e-01,
        -3.3298e+00,  1.6525e+00, -7.7475e-01,  1.8055e+00,  2.4089e+00,
        -5.4255e-02, -2.0334e+00,  5.7014e-01, -4.7004e+00,  9.0794e-01,
        -3.2066e-01, -6.5802e-01,  1.8745e+00, -1.8683e+00,  5.5586e-01,
        -2.8878e-01,  5.2947e-01,  1.3586e+00,  1.6520e+00,  5.5001e-01,
         1.0854e+00,  2.6800e-01, -1.5278e+00, -1.6991e+00,  2.6149e-01,
         1.9350e-01, -2.1655e+00,  2.3297e+00, -5.6678e-01,  2.6218e+00,
         3.1062e+00,  2.8352e-01, -1.5966e+00, -2.2085e+00,  4.1520e-01,
        -8.3689e-01,  2.0700e+00,  3.8966e-01,  2.0180e-01,  6.3099e-01,
         2.9767e-01, -1.3401e+00, -2.0358e-01,  8.2342e-01, -3.0265e-01,
        -1.3500e+00,  1.8864e+00,  2.5154e+00, -1.7893e+00,  7.5290e-01,
         1.3409e+00,  3.1824e-01, -1.8128e+00, -1.2409e+00, -1.5763e-03,
        -2.0357e+00,  8.1066e-01, -1.4737e+00,  3.4677e-02,  7.3462e-01,
        -5.1422e-01,  5.4877e-01, -9.8326e-01, -2.7177e+00,  2.3824e+00,
        -6.4974e-02,  8.0561e-01,  1.5707e+00, -2.9389e+00,  2.0913e+00,
         1.3360e+00,  2.3601e+00,  2.2581e+00, -3.9577e-01,  2.1458e+00,
        -1.6210e+00,  2.7131e+00, -1.0139e-01], device='cuda:0'),
    expected=tensor([-9.8341e-01,  2.2043e+00, -8.5609e-01,  1.0821e-01, -1.5677e+00,
         1.2017e+00, -3.2008e-01, -5.4052e-01, -2.4166e-01,  1.4987e+00,
         9.0678e-01, -2.5401e+00,  1.0601e+00,  4.9114e-01,  9.3326e-01,
        -2.0346e+00,  9.1439e-04, -1.6040e+00,  2.1894e+00,  1.0840e+00,
         2.8241e+00, -8.8914e-01, -6.2965e-01,  7.6015e-01, -2.6187e-01,
        -1.0823e+00, -1.9986e+00,  4.2247e-01, -1.9694e+00,  1.9601e-01,
        -1.4765e+00,  8.4396e-01,  3.0701e-01,  1.5813e-01, -8.6276e-01,
         1.2773e+00, -7.9689e-01,  3.2489e+00, -6.1513e-02,  3.4948e+00,
        -1.1431e+00, -1.7335e+00,  7.7941e-01, -1.6171e+00, -2.1219e+00,
         1.4460e+00,  3.0477e+00, -5.2305e-01, -3.3852e-01,  3.3611e-01,
        -3.3298e+00,  1.6525e+00, -7.7475e-01,  1.8055e+00,  2.4089e+00,
        -5.4255e-02, -2.0334e+00,  5.7014e-01, -4.7004e+00,  9.0794e-01,
        -3.2066e-01, -6.5802e-01,  1.8745e+00, -1.8683e+00,  5.5586e-01,
        -2.8878e-01,  5.2947e-01,  1.3586e+00,  1.6520e+00,  5.5001e-01,
         1.0854e+00,  2.6800e-01, -1.5278e+00, -1.6991e+00,  2.6149e-01,
         1.9350e-01, -2.1655e+00,  2.3297e+00, -5.6678e-01,  2.6218e+00,
         3.1062e+00,  2.8352e-01, -1.5966e+00, -2.2085e+00,  4.1520e-01,
        -8.3689e-01,  2.0700e+00,  3.8966e-01,  2.0180e-01,  6.3099e-01,
         2.9767e-01, -1.3401e+00, -2.0358e-01,  8.2342e-01, -3.0265e-01,
        -1.3500e+00,  1.8864e+00,  2.5154e+00, -1.7893e+00,  7.5290e-01,
         1.3409e+00,  3.1824e-01, -1.8128e+00, -1.2409e+00, -1.5763e-03,
        -2.0357e+00,  8.1066e-01, -1.4737e+00,  3.4677e-02,  7.3462e-01,
        -5.1422e-01,  5.4877e-01, -9.8326e-01, -2.7177e+00,  2.3824e+00,
        -6.4974e-02,  8.0561e-01,  1.5707e+00, -2.9389e+00,  2.0913e+00,
         1.3360e+00,  2.3601e+00,  2.2581e+00, -3.9577e-01,  2.1458e+00,
        -1.6210e+00,  2.7131e+00, -1.0139e-01], device='cuda:0'),
    rtol=1.3e-06,
    atol=1e-05,
    equal_nan=True,
    check_device=False,
    check_dtype=True,
    check_layout=False,
    check_stride=False,
)

resulted in the unexpected exception above. If you are a user and see this message during normal operation please file an issue at https://github.com/pytorch/pytorch/issues. If you are a developer and working on the comparison functions, please except the previous error and raise an expressive `ErrorMeta` instead.

----------------------------------------------------------------------
Ran 4 tests in 17.795s

**Comment by EikanWang** [comment]
@fbgheith , may I know if the failure was captured by Meta's internal CI?

**Comment by atalman** [comment]
@atalman has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59399546).

**Review by atalman — APPROVED** [review]
lgtm. internal diff is green

**Pull request closed without merging**